attivita/migrations/0018_negazione_automatica_vecchie.py

```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from datetime import datetime, date, timedelta
+
+from django.db import migrations, models
+
+
+def nega_partecipazioni_vecchie(apps, schema_editor):
+
+    Partecipazione = apps.get_model("attivita", "Partecipazione")
+    Autorizzazione = apps.get_model("base", "Autorizzazione")
+
+    partecipzioni_pendenti = Partecipazione.objects.filter(confermata=False, ritirata=False, automatica=False)
+    totale = partecipzioni_pendenti.count()
+    print("  => 0018 trovate %d partecipazioni di attivita pendenti" % (totale))
+
+    autorizzazioni_con_scadenza = 0
+    autorizzazioni_senza_scadenza = 0
+    for partecipazione in partecipzioni_pendenti:
+        # Autorizzazioni senza esito della partecipazione pendente
+        autorizzazioni_partecipazione = Autorizzazione.objects.filter(oggetto_id=partecipazione.pk, concessa=None)
+        for autorizzazione in autorizzazioni_partecipazione:
```

Comment:
Questo ha complessita' molto alta, ci sono moltissime partecipazioni pendenti (10k-100k), e questo codice effettua una query per ognuna.

Non potremmo fare tutto in due semplicissime query con `.update`? Qualcosa come:
```python
autorizzazioni = \
  Autorizzazione.objects.filter(oggetto__confermata=False, ritirata=False, automatica=False)
                        .filter(scadenza__isnull=False)
                        .exclude(tipo_gestione__"N")
autorizzazioni.update(tipo_gestione="N")
```

Stiamo parlando di 1-5 query contro >10k!

Reply:
Grazie della review. Visto il numero richieste pendenti capisco l'esigenza di non fare una moltitudine di query. Non ho capito bene il suggerimento che mi hai dato in quanto il modello Autorizzazione non prevede campi come ritirata (presente in attività Partecipazione) o oggetto_confermata.
Forse una soluzione più semplice (credo intedessi questo) potrebbe essere di considerare solo il modello Autorizzazione, senza considerare la data di scadenza.

`autorizzazioni = Autorizzazione.objects.filter(concessa__isnull=True, destinatario_ruolo=INCARICO_GESTIONE_ATTIVITA_PARTECIPANTI).exclude(tipo_gestione="N")`

In questo modo si prendono solo le autorizzazioni per le attività di partecipazione (destinatario_ruolo è uguale a 'ATT-PART'), senza esito (concessa null) e con il campo tipo gestione non impostato correttamente su negazione automatica (che attualmente per le richieste di partecipazione pendenti è impostato di default su 'N').

Infine facciamo l'update con la data attuale che poi nella pratica scade subito dopo.

`autorizzazioni.update(scadenza=datetime.now(), tipo_gestione="N")`

La mia soluzione voleva mantenere una coerenza sulla data di scadenza a 30 giorni, pensando che le vecchie autorizzazioni potessero avere impostato o meno (null) un valore per la scadenza. Volendo impostare una data di scadenza non vedevo altro modo se non iterare (almeno una volta) e prendere il valore del campo creazione di ciascuna autorizzazione per poi aggiungerci 30 giorni.


Se ritieni corretta l'idea di non considerare una data di scadenza particolare procederei con le modifiche proposte:

`autorizzazioni = Autorizzazione.objects.filter(concessa__isnull=True, destinatario_ruolo=INCARICO_GESTIONE_ATTIVITA_PARTECIPANTI).exclude(tipo_gestione="N")
autorizzazioni.update(scadenza=datetime.now(), tipo_gestione="N")`